kv_backend.cc

```diff
@@ -1,33 +1,71 @@
-#include <string>
-#include <map>
-#include "key_value_interface.cc"
-
-// true kv_backend implementation that stores all raw data
-class KeyValueBackEnd : public KeyValueInterface{
-	public:
-		KeyValueBackEnd();
-		~KeyValueBackEnd();
-		std::string Get(const std::string& key);
-		std::string Put(const std::string& key, const std::string& chirp_value);
-		std::string DeleteKey(const std::string& key);
-	private:
-		// key - std::string, value - serialized chirp or user object
-		std::map<std::string, std::string> chirp_map_;	
-};
-
-std::string KeyValueBackEnd::Get(const std::string& key) {
-	return chirp_map_.at(key);
+#include "kv_backend.h"
+
+#include <iostream>
+
+// global mutex that keeps the database thread safe
+std::mutex mutex_lock;
+
+KeyValueBackEnd::KeyValueBackEnd() {
+	// initialize the key counter to 0 when the database is created
+	std::pair<std::string, std::vector<std::string> > key_counter;
+	std::vector<std::string> v;
+	v.push_back("0");
+	key_counter.first = "keyCounterForDataMap";
+	key_counter.second = v;
+	data_map_.insert(key_counter);
+	// create newest and seen keys
+	std::pair<std::string, std::vector<std::string> > newest_pair;
+	std::vector<std::string> vector_newest;
+	newest_pair.first = "newest";
+	newest_pair.second = vector_newest;
+	data_map_.insert(newest_pair);
+	std::pair<std::string, std::vector<std::string> > seen_pair;
+	std::vector<std::string> vector_seen;
+	seen_pair.first = "seen";
+	seen_pair.second = vector_seen;
+	data_map_.insert(seen_pair);
+}
+
+KeyValueBackEnd::~KeyValueBackEnd() {}
+
+std::vector<std::string> KeyValueBackEnd::Get(const std::string& key) {
+	std::lock_guard<std::mutex> lock(mutex_lock);
+	return data_map_.at(key);
+}
+
+void KeyValueBackEnd::Put(const std::string& key, const std::string& value) {
+	std::lock_guard<std::mutex> lock(mutex_lock);
+	std::map<std::string,std::vector<std::string> >::iterator it = data_map_.find(key);
+	// check to see if we are updating either of our special key's:
+	// "newest" which supports monitor or "keyCounterForDataMap" which ensures unique keys
+	if ((key == "keyCounterForDataMap")) {
```

Comment:
the kv data structure shouldn't know anything about what it is storing, this should be handled by the service layer